### PR TITLE
Additional functions, EXPLODE, and expressions in subfields

### PR DESCRIPTION
**Pull request by 0aix**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by 0aix** [comment]
I feel things are a bit messy and can be done better so looking for comments on approach.

**Comment by 0aix** [comment]
ready for review

**Review by akshaisarma — COMMENTED** [review]
I need to think a bit about how lateral view and explode are modeled. Wanted to submit some comments in the meantime

File: src/main/java/com/yahoo/bullet/querying/Querier.java [hunk 1]
```diff
@@ -356,14 +361,21 @@ private void start() {
         }
 
         Query query = runningQuery.getQuery();
+
         Expression filter = query.getFilter();
         if (filter != null) {
             this.filter = new Filter(filter);
         }
 
+        TableFunction tableFunction = query.getTableFunction();
```
[the remark this answers is not in the document]
Comment (0aix) [reply src/main/java/com/yahoo/bullet/querying/Querier.java:370]: I know what you mean by writing an IdentityTableFunction, but I'm not sure about the pulling out and exposing the TableFunctor. Do you mean checking if it's null is undesirable? I feel like we do something similar with the Filter above and the Projection below though.

File: src/main/java/com/yahoo/bullet/querying/tablefunctors/TableFunctor.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * Table functors are built from table functions. A table functor is applied to a {@link BulletRecord} and generates a
+ * list of {@link BulletRecord}.
+ *
+ * Classes inheriting from this class must implement {@link TableFunctor#apply(BulletRecord, BulletRecordProvider)} but
+ * not the lateral view and outer options which are already implemented in {@link TableFunctor#getRecords(BulletRecord,
+ * BulletRecordProvider)}.
+ */
+@AllArgsConstructor
+public abstract class TableFunctor implements Serializable {
```
[the remark this answers is not in the document]
Comment (0aix) [reply src/main/java/com/yahoo/bullet/querying/tablefunctors/TableFunctor.java:27]: We can. I was thinking Expression -> Evaluator so TableFunction -> TableFunctor. If I rename the package to tablefunctions, should I rename TableFunctor to TableFunction?

File: src/main/java/com/yahoo/bullet/querying/evaluators/NAryOperations.java [hunk 3]
```diff
@@ -60,4 +68,115 @@ static TypedObject ternary(List<Evaluator> evaluators, BulletRecord record) {
...
+        }
+        int length = ((Number) lengthArg.getValue()).intValue();
+        if (length <= 0) {
+            return TypedObject.valueOf("");
+        }
+        int end = Math.min(start + length, string.length());
+        return TypedObject.valueOf(string.substring(start, end));
+    }
+
+    // TODO pr: should this use UTC or system timezone?
```
[the remark this answers is not in the document]
Comment (0aix) [reply src/main/java/com/yahoo/bullet/querying/evaluators/NAryOperations.java:153]: So the concern when choosing one or the other is that hive actually does the system timezone, and a couple of the validatar tests expect that behavior. But, I just realized we could manually correct the tests to the right timezone with some simple math if we know it's UTC, so we should be all good to use UTC. 👍

File: src/main/java/com/yahoo/bullet/querying/tablefunctors/OuterableTableFunctor.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    @Override
+    public List<BulletRecord> apply(BulletRecord record, BulletRecordProvider provider) {
+        List<BulletRecord> records = outerableApply(record, provider);
+        if (records.isEmpty() && outer) {
+            return Collections.singletonList(provider.getInstance());
+        }
+        return records;
+    }
+
+    protected abstract List<BulletRecord> outerableApply(BulletRecord record, BulletRecordProvider provider);
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/querying/tablefunctors/OuterableTableFunctor.java:37]: Javadoc

**Review by akshaisarma — DISMISSED** [review]

**Review by akshaisarma — APPROVED** [review]

**Pull request merged**